domains/MatchRoom.js

```diff
@@ -257,93 +267,132 @@ class MatchRoom extends Room {
 		return this.state.players.find(player => player.id === player_id);
 	}
 
-	onAdminMessage(message) {
-		const [command, ...args] = message;
-		let forward_to_views = true;
-		let update_admin = true;
+	getMaxPossiblePlayers() {
+		return Math.min(this.getViewMeta().players || Infinity, MAX_PLAYERS);
+	}
 
-		// TODO: Extract this encompassing try..catch to own method
-		try {
-			switch (command) {
-				case 'getState': {
-					forward_to_views = false;
-					break;
-				}
+	initAutoJoin() {
+		// 1. drop all players
+		for (let idx = this.state.players.length; idx--; ) {
+			if (this.state.players[idx].id) {
+				this.setPlayer(idx, null);
+			}
+		}
 
-				case 'setPlayer': {
-					const [p_num, p_id] = args;
-					console.log('setPlayer()', p_id, typeof p_id);
+		// 1. add players in order of arrival
+		const sorted_producers = [...this.producers].sort(
+			(u1, u2) => u2.match_room_join_ts - u1.match_room_join_ts // breaks encapsulation T_T
+		);
 
-					let player_data;
-					let player_id = `${p_id}`;
+		// 3. fill as much players as the layout allows
+		for (
+			let idx = 0;
+			idx < Math.min(this.getMaxPossiblePlayers(), this.producers.size);
+			idx++
+		) {
+			this.setPlayer(idx, sorted_producers[idx].id);
+		}
+	}
 
-					this.assertValidPlayer(p_num);
+	autoJoinUser(user) {
+		for (let idx = 0; idx < this.getMaxPossiblePlayers(); idx++) {
+			if (!this.state.players[idx]?.id) {
+				this.setPlayer(idx, user.id);
+				this.sendStateToAdmin();
+				break;
+			}
+		}
+	}
 
-					const old_player_id = this.state.players[p_num].id;
+	setPlayer(player_num, p_id) {
+		console.log('setPlayer()', p_id, typeof p_id);
 
-					if (old_player_id && player_id != old_player_id) {
-						// player at p_num is being replaced
-						// check if old player was used in more than one slot
-						// if not, then user needs to be informed it is dropped as a player
-						const still_around =
-							this.state.players.filter(player => player.id === old_player_id)
-								.length > 1;
+		let player_data;
+		const player_id = `${p_id}`;
 
-						if (!still_around) {
-							const user = this.getProducer(this.state.players[p_num].id);
+		this.assertValidPlayer(p_num);
 
-							if (user) {
-								// hmm, is this necessary?
-								user.getProducer().send(['dropPlayer']);
-							}
-						}
-					}
+		const old_player_id = this.state.players[p_num].id;
 
-					const user = this.getProducer(player_id);
+		if (old_player_id && player_id != old_player_id) {
+			// player at p_num is being replaced
+			// check if old player was used in more than one slot
+			// if not, then user needs to be informed it is dropped as a player
+			const still_around =
```

Comment:
This seems to hint at a problem when swapping player slots. It will only work for a part of the players without dropping them in between when done for each player after each other. There would be higher flexibility of you can pass a list of slots and player to this function. Than at begin of the function make a copy of playerdata of all slots so no one gets lost if he just swaps place with someone else.

Reply:
Thanks for the comment, but I'm afraid I do not understand this line 😅 

> It will only work for a part of the players without dropping them in between when done for each player after each other.

In any case, this is not new code, it's the existing code for assigning a producer to a player. It is driven by the admin UI, and there is no swap function. In the admin UI, a swap is a 2 step process where the same user is assigned to 2 players, before the second user is assigned to the other desired player.

I'm thinking it'll be OK for now 🤞, and I can revisit that later.